### ISSUE TITLE
[java] Should the hasMoreElements/hasNext/hasPrevious preceding nextElement/next/previous/ protocol be enforced via a rule?

### ISSUE THREAD

**Issue by Fernal73** [pmd/pmd#2875]
**Description:**

Should the **hasMoreElements/hasNext/hasPrevious** preceding **nextElement/next/previous** protocol be enforced via a rule?

Sonar has an existing [rule](https://rules.sonarsource.com/java/RSPEC-1849) that ensures that `hasNext`  implementation does not invoke `next`.

This rule can generate false positives when `has*` is not invoked prior to `next*/previous` when calling next or previous would be legitimate.
(I'll leave code examples supporting this as an exercise for you. A good bet would be StackOverflow where hasNext/next is discussed.)

Would it be a better bet to have a rule that every call to `iterator.remove` is preceded by `next` or `previous` since [remove is allowed only once per next or previous invocation](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Iterator.html#remove())?

**Comment by Fernal73** [comment]
The `StringTokenizer` class and its methods hasMoreTokens/nextToken could be added to this rule.

While the [StringTokenizer class is considered a legacy class](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/StringTokenizer.html) and String.split method is recommended instead, there is a negative performance impact from using a regex pattern to tokenize strings.

https://www.java67.com/2019/12/difference-between-stringtokenizer-and.html

https://stackoverflow.com/a/691224/3924108

https://www.javamex.com/tutorials/regular_expressions/splitting_tokenisation_performance.shtml

Can we have another rule that deprecates the use of StringTokenizer in new code as long as any performance penalty is ignored? Or do we disallow StringTokenizer [when the default space separator is specified i.e, default constructor](http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/java/lang/String.java)?

**Comment by Fernal73** [comment]
Running StringTokenizer and String.split through jmh gives the following results (for SPACE separator):

```
Benchmark                                                          Mode      Cnt      Score      Error   Units
Tokenizer.measureStringSplit                                      thrpt       25      0.114 ±    0.001  ops/us
Tokenizer.measureStringTokenizer                                  thrpt       25      0.077 ±    0.001  ops/us
Tokenizer.measureStringSplit                                       avgt       25      8.835 ±    0.041   us/op
Tokenizer.measureStringTokenizer                                   avgt       25     12.909 ±    0.084   us/op
```

**Comment by Fernal73** [comment]
Updated the above jmh performance stats:

```java
package jmh;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.StringTokenizer;
import java.util.concurrent.TimeUnit;
import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.BenchmarkMode;
import org.openjdk.jmh.annotations.Level;
import org.openjdk.jmh.annotations.Mode;
import org.openjdk.jmh.annotations.OutputTimeUnit;
import org.openjdk.jmh.annotations.Scope;
import org.openjdk.jmh.annotations.Setup;
import org.openjdk.jmh.annotations.State;
import org.openjdk.jmh.annotations.TearDown;
import org.openjdk.jmh.runner.Runner;
import org.openjdk.jmh.runner.RunnerException;
import org.openjdk.jmh.runner.options.Options;
import org.openjdk.jmh.runner.options.OptionsBuilder;

@State(Scope.Thread)
public class Tokenizer {

  String sample;
  int tokenCount;
  int splitCount;
  int splitRegexCount;

  @Setup(Level.Trial)
  public void doSetup() {
    StringBuilder sb = new StringBuilder();
    for (int i = 100_000; i < 100_000 + 60; i++) {
      sb.append(i).append(' ');
      if (i % 5 == 0)
        sb.append("  ");
    }
    sample = sb.toString();
  }

  @TearDown(Level.Trial)
  public void doTearDown() {
    sample = null;
    if (tokenCount > 0)
      System.out.println("Token count: " + tokenCount);
    if (splitCount > 0)
      System.out.println("Split count: " + splitCount);
    if (splitRegexCount > 0)
      System.out.println("Split regex count: " + splitRegexCount);
  }

  @Benchmark
  @BenchmarkMode(Mode.AverageTime)
  @OutputTimeUnit(TimeUnit.MICROSECONDS)
  public void measureStringTokenizer(Tokenizer t) {
    StringTokenizer st = new StringTokenizer(t.sample, " ");
    List<String> list = new ArrayList<>();
    while (st.hasMoreTokens())
      list.add(st.nextToken());
    t.tokenCount = list.size();
  }

  @Benchmark
  @BenchmarkMode(Mode.AverageTime)
  @OutputTimeUnit(TimeUnit.MICROSECONDS)
  public void measureStringSplitRegex(Tokenizer t) {
    String[] tokens = t.sample.split("\\s+");
    List<String> list = Arrays.asList(tokens);
    t.splitRegexCount = list.size();
  }

  @Benchmark
  @BenchmarkMode(Mode.AverageTime)
  @OutputTimeUnit(TimeUnit.MICROSECONDS)
  public void measureStringSplit(Tokenizer t) {
    String[] tokens = t.sample.split(" ");
    List<String> list = Arrays.asList(tokens);
    t.splitCount = list.size();
  }

  public static void main(String[] args) throws RunnerException {
    Options opt = new OptionsBuilder()
                      .include(Tokenizer.class.getSimpleName())
                      .forks(1)
                      .build();
    new Runner(opt).run();
  }
}
```

Source above: Let me know how it could be enhanced.

**Comment by Fernal73** [comment]
> Can we have another rule that deprecates the use of StringTokenizer in new code as long as any performance penalty is ignored?

It's apparent that there's no easy way to identify a migration path from **StringTokenizer** to **String.split** , i.e., through a parser, since the input string plays an equally important role as the delimiter(s) or regex(es) in identifying behaviour i.e., tokens generated. 

It might be best to have a suggestive rule that prescribes **String.split* when **StringTokenizer** class is encountered which (like the use of keyword `volatile`) can be ignored if you know what you're doing.

**Comment by oowekyala** [comment]
> It might be best to have a suggestive rule that prescribes String.split when StringTokenizer class is encountered which (like the use of keyword volatile) can be ignored or suppressed when you know what you're doing.

This is unrelated to the original issue, which is about having a rule that enforces that `next` calls and such are dominated by `hasNext` checks. Please open another issue with your analysis of StringTokenizer vs String.split

**Comment by Fernal73** [comment]
> Would it be better to have a rule that every call to `iterator.remove` is preceded by **`next`** or **`previous`** since [remove is allowed only once per next or previous invocation](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Iterator.html#remove())?

I've updated the original post for rule name and category. I think all three elements: ** hasNext, Sonar rule and remove  ** can be bundled in one rule with properties to enable or disable each aspect (if needed).
